File: gh-sync/src/arguments/parser.rs
use std::{
    env::args,
    io::{StdoutLock, Write, stdout},
    process::ExitCode,
    string::String,
    vec::Vec,
};

use super::{print_help::print_help_message, print_version::print_version_number};

use crate::github::synchronize_repositories::synch_repo;

// Command Line Argument Parser
pub fn parse() -> ExitCode {
    let command_line_arguments: Vec<String> = args().collect();
    let mut standard_output: StdoutLock = stdout().lock();

    if command_line_arguments.len() != 2 {
        writeln!(
            standard_output,
            "\x1b[31;1;3;4mCommand or Flag Required but not Both:\x1b[0m {:#?}",
            command_line_arguments
        )
        .unwrap();
        writeln!(standard_output, "").unwrap();
        print_help_message();
        eprintln!("\x1b[31;1;3;4mError(1) - Exiting GitHub Syncrhonize\x1b[0m");
        return ExitCode::FAILURE;
    } else {
        match command_line_arguments[1].trim() {
            "help" | "--h" => {
                print_help_message();
            }
            "version" | "--v" => {
                print_version_number();
            }
            &_ => {
                synch_repo(command_line_arguments[1].trim());
            }
        };
    };

    return ExitCode::SUCCESS;
}

File: null-log/src/arguments/parser.cc
#include "./parser.hh"
#include "../null/cleaner.hh"
#include "./help_message.hh"
#include "./version_message.hh"

#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Parse Command Line Arguments
uint8_t parse_arguments(size_t length, char *arguments[]) {
  int8_t index;

  if (length < 1 || length > 1) {
    printf("\x1b[31;1;3;4mCommand or Flag Required but not Both:\x1b[0m\n");
    for (index = 0; index < length; index = index + 1) {
      printf("%s\n", arguments[index]);
    }
    printf("");
    print_help();
    printf("\x1b[31;1;3;4mError(1) - Exiting Null Log\x1b[0m");
    exit(1);
  } else if (strcmp(arguments[1], "help")) {
    print_help();
  } else if (strcmp(arguments[1], "--h")) {
    print_help();
  } else if (strcmp(arguments[1], "version")) {
    print_version();
  } else if (strcmp(arguments[1], "--v")) {
    print_version();
  } else {
    clean_logs(arguments[1]);
  }

  return 0;
}
